File: ios/Wrappers/AuthParamsWrapper.swift
//
//  AuthParamsWrapper.swift
//  XMTPReactNative
//
//  Created by Naomi Plasterer on 6/24/24.
//

import Foundation
import XMTP


struct AuthParamsWrapper {
	let environment: String
	let dbDirectory: String?
	let historySyncUrl: String?
	let customLocalUrl: String?
	let deviceSyncEnabled: Bool
	let debugEventsEnabled: Bool
	let appVersion: String?
	let gatewayHost: String?
	let forkRecoveryOptions: ForkRecoveryOptions?

	init(
		environment: String, dbDirectory: String?,
		historySyncUrl: String?, customLocalUrl: String?,
		deviceSyncEnabled: Bool, debugEventsEnabled: Bool,
		appVersion: String?, gatewayHost: String?,
		forkRecoveryOptions: ForkRecoveryOptions?
	) {
		self.environment = environment
		self.dbDirectory = dbDirectory
		self.historySyncUrl = historySyncUrl
		self.customLocalUrl = customLocalUrl
		self.deviceSyncEnabled = deviceSyncEnabled
		self.debugEventsEnabled = debugEventsEnabled
		self.appVersion = appVersion
		self.gatewayHost = gatewayHost
		self.forkRecoveryOptions = forkRecoveryOptions
	}

	private static func createForkRecoveryOptions(
		enableRecoveryRequestsString: String?,
		groupsToRequestRecovery: [String]?,
		disableRecoveryResponse: Bool?,
		workerIntervalNs: UInt64?
	) -> ForkRecoveryOptions? {
		// If none of the fork recovery options are provided, return nil
		if enableRecoveryRequestsString == nil && 
		   groupsToRequestRecovery == nil && 
		   disableRecoveryResponse == nil && 
		   workerIntervalNs == nil {
			return nil
		}
		
		return ForkRecoveryOptions(
			enableRecoveryRequests: convertToForkRecoveryPolicy(enableRecoveryRequestsString),
			groupsToRequestRecovery: groupsToRequestRecovery ?? [],
			disableRecoveryResponses: disableRecoveryResponse,
			workerIntervalNs: workerIntervalNs
		)
	}
	
	private static func convertToForkRecoveryPolicy(_ enableRecoveryRequestsString: String?) -> ForkRecoveryPolicy {
		switch enableRecoveryRequestsString {
		case "none":
			return .none
		case "all":
			return .all
		case "groups":
			return .allowlistedGroups
		default:
			return .none
		}
	}

	static func authParamsFromJson(_ authParams: String) -> AuthParamsWrapper {
		guard let data = authParams.data(using: .utf8),
		      let jsonOptions = try? JSONSerialization.jsonObject(
		      	with: data, options: []
		      ) as? [String: Any]
		else {
			return AuthParamsWrapper(
				environment: "dev", dbDirectory: nil,
				historySyncUrl: nil, customLocalUrl: nil,
				deviceSyncEnabled: true, debugEventsEnabled: false,
				appVersion: nil, gatewayHost: nil,
				forkRecoveryOptions: nil
			)
		}

		let environment = jsonOptions["environment"] as? String ?? "dev"
		let dbDirectory = jsonOptions["dbDirectory"] as? String
		let historySyncUrl = jsonOptions["historySyncUrl"] as? String
		if let historySyncUrl = historySyncUrl {
			setenv("XMTP_HISTORY_SERVER_ADDRESS", historySyncUrl, 1)
		}
		let customLocalUrl = jsonOptions["customLocalUrl"] as? String
		if let customLocalUrl = customLocalUrl {
			setenv("XMTP_NODE_ADDRESS", customLocalUrl, 1)
		}
		let deviceSyncEnabled =
			jsonOptions["deviceSyncEnabled"] as? Bool ?? true
		let debugEventsEnabled =
			jsonOptions["debugEventsEnabled"] as? Bool ?? false
		let appVersion = jsonOptions["appVersion"] as? String
		let gatewayHost = jsonOptions["gatewayHost"] as? String
		
		// Parse fork recovery options
		let enableRecoveryRequestsString = jsonOptions["enableRecoveryRequests"] as? String
		let groupsToRequestRecovery = jsonOptions["groupsToRequestRecovery"] as? [String]
		let disableRecoveryResponse = jsonOptions["disableRecoveryResponses"] as? Bool
		let workerIntervalNs = jsonOptions["workerIntervalNs"] as? UInt64
		
		let forkRecoveryOptions = createForkRecoveryOptions(
			enableRecoveryRequestsString: enableRecoveryRequestsString,
			groupsToRequestRecovery: groupsToRequestRecovery,
			disableRecoveryResponse: disableRecoveryResponse,
			workerIntervalNs: workerIntervalNs
		)

		return AuthParamsWrapper(
			environment: environment,
			dbDirectory: dbDirectory,
			historySyncUrl: historySyncUrl,
			customLocalUrl: customLocalUrl,
			deviceSyncEnabled: deviceSyncEnabled,
			debugEventsEnabled: debugEventsEnabled,
			appVersion: appVersion,
			gatewayHost: gatewayHost,
			forkRecoveryOptions: forkRecoveryOptions
		)
	}
}

struct WalletParamsWrapper {
	let signerType: SignerType
	let chainId: Int64?
	let blockNumber: Int64?

	init(signerType: SignerType, chainId: Int64?, blockNumber: Int64?) {
		self.signerType = signerType
		self.chainId = chainId
		self.blockNumber = blockNumber
	}

	static func walletParamsFromJson(_ walletParams: String)
		-> WalletParamsWrapper
	{
		guard let data = walletParams.data(using: .utf8),
		      let jsonOptions = try? JSONSerialization.jsonObject(
		      	with: data, options: []
		      ) as? [String: Any]
		else {
			return WalletParamsWrapper(
				signerType: SignerType.EOA, chainId: nil, blockNumber: nil
			)
		}

		let walletTypeString = jsonOptions["signerType"] as? String ?? "EOA"
		let chainId = jsonOptions["chainId"] as? Int64
		let blockNumber = jsonOptions["blockNumber"] as? Int64

		let signerType = {
			switch walletTypeString {
			case "SCW":
				return SignerType.SCW
			default:
				return SignerType.EOA
			}
		}()

		return WalletParamsWrapper(
			signerType: signerType,
			chainId: chainId,
			blockNumber: blockNumber
		)
	}
}

File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/AuthParamsWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.JsonArray
import com.google.gson.JsonParser
import org.xmtp.android.library.ForkRecoveryOptions
import org.xmtp.android.library.ForkRecoveryPolicy
import org.xmtp.android.library.SignerType
import java.math.BigInteger

class AuthParamsWrapper(
    val environment: String,
    val dbDirectory: String?,
    val historySyncUrl: String?,
    val customLocalUrl: String?,
    val deviceSyncEnabled: Boolean,
    val debugEventsEnabled: Boolean,
    val appVersion: String?,
    val gatewayHost: String?,
    val forkRecoveryOptions: ForkRecoveryOptions?
) {
    companion object {
        private fun createForkRecoveryOptions(
            enableRecoveryRequestsString: String?,
            groupsToRequestRecovery: JsonArray?,
            disableRecoveryResponse: Boolean?,
            workerIntervalNs: BigInteger?
        ): ForkRecoveryOptions? {
            // If none of the fork recovery options are provided, return null
            if (enableRecoveryRequestsString == null && 
                groupsToRequestRecovery == null && 
                disableRecoveryResponse == null && 
                workerIntervalNs == null) {
                return null
            }

            // Convert groupsToRequestRecovery JsonArray to List<String>
            val groupsList = groupsToRequestRecovery?.map { it.asString } ?: emptyList()

            //convert BigInt to ULong safely
            val workerIntervalNsLong = workerIntervalNs?.let { bigInt ->
                try {
                    if (bigInt <= BigInteger.valueOf(Long.MAX_VALUE)) {
                        bigInt.toLong().toULong()
                    } else {
                        ULong.MAX_VALUE // Clamp to max value if too large
                    }
                } catch (e: Exception) {
                    null
                }
            }

            return ForkRecoveryOptions(
                enableRecoveryRequests = convertToForkRecoveryPolicy(enableRecoveryRequestsString),
                groupsToRequestRecovery = groupsList,
                disableRecoveryResponses = disableRecoveryResponse,
                workerIntervalNs = workerIntervalNsLong
            )
        }

        private fun convertToForkRecoveryPolicy(enableRecoveryRequestsString: String?): ForkRecoveryPolicy {
            return when (enableRecoveryRequestsString) {
                "none" -> ForkRecoveryPolicy.None
                "all" -> ForkRecoveryPolicy.All
                "groups" -> ForkRecoveryPolicy.AllowlistedGroups
                else -> ForkRecoveryPolicy.None
            }
        }

        fun authParamsFromJson(authParams: String): AuthParamsWrapper {
            val jsonOptions = JsonParser.parseString(authParams).asJsonObject
            val enableRecoveryRequestsString = if (jsonOptions.has("enableRecoveryRequests")) jsonOptions.get("enableRecoveryRequests").asString else null
            val groupsToRequestRecovery = if (jsonOptions.has("groupsToRequestRecovery")) jsonOptions.get("groupsToRequestRecovery").asJsonArray else null
            val disableRecoveryResponse = if (jsonOptions.has("disableRecoveryResponses")) jsonOptions.get("disableRecoveryResponses").asBoolean else null
            val workerIntervalNs = if (jsonOptions.has("workerIntervalNs")) jsonOptions.get("workerIntervalNs").asBigInteger else null

            val forkRecoveryOptions = createForkRecoveryOptions(
                enableRecoveryRequestsString,
                groupsToRequestRecovery,
                disableRecoveryResponse,
                workerIntervalNs
            )

            return AuthParamsWrapper(
                jsonOptions.get("environment").asString,
                if (jsonOptions.has("dbDirectory")) jsonOptions.get("dbDirectory").asString else null,
                if (jsonOptions.has("historySyncUrl")) jsonOptions.get("historySyncUrl").asString else null,
                if (jsonOptions.has("customLocalUrl")) jsonOptions.get("customLocalUrl").asString else null,
                if (jsonOptions.has("deviceSyncEnabled")) jsonOptions.get("deviceSyncEnabled").asBoolean else true,
                if (jsonOptions.has("debugEventsEnabled")) jsonOptions.get("debugEventsEnabled").asBoolean else false,
                if (jsonOptions.has("appVersion")) jsonOptions.get("appVersion").asString else null,
                if (jsonOptions.has("gatewayHost")) jsonOptions.get("gatewayHost").asString else null,
                forkRecoveryOptions
            )
        }
    }
}

class WalletParamsWrapper(
    val signerType: SignerType = SignerType.EOA,
    val chainId: Long?,
    val blockNumber: Long?,
) {
    companion object {
        fun walletParamsFromJson(walletParams: String): WalletParamsWrapper {
            val jsonOptions = JsonParser.parseString(walletParams).asJsonObject
            return WalletParamsWrapper(
                if (jsonOptions.has("signerType")) {
                    when (jsonOptions.get("signerType").asString) {
                        "SCW" -> SignerType.SCW
                        else -> SignerType.EOA
                    }
                } else {
                    SignerType.EOA
                },
                if (jsonOptions.has("chainId")) jsonOptions.get("chainId").asLong else null,
                if (jsonOptions.has("blockNumber")) jsonOptions.get("blockNumber").asLong else null,
            )
        }
    }
}

